File: app/src/main/java/com/checkin/app/checkin/Auth/SignupUserInfoFragment.java
package com.checkin.app.checkin.Auth;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;

import com.checkin.app.checkin.Misc.DebouncedOnClickListener;
import com.checkin.app.checkin.R;
import com.checkin.app.checkin.User.UserModel.GENDER;
import com.fasterxml.jackson.databind.JsonNode;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;


public class SignupUserInfoFragment extends Fragment {
    public static final String KEY_NAME = "name";
    @BindView(R.id.ed_firstname)
    EditText edFirstName;
    @BindView(R.id.ed_lastname)
    EditText edLastName;
    @BindView(R.id.im_male)
    FrameLayout imMale;
    @BindView(R.id.im_female)
    FrameLayout imFemale;
    @BindView(R.id.ed_Username)
    EditText edUsername;
    @BindView(R.id.btn_enter)
    Button btnEnter;

    private AuthFragmentInteraction fragmentInteraction;
    private Unbinder unbinder;
    private GENDER genderChosen = null;

    public SignupUserInfoFragment() {
    }

    public static SignupUserInfoFragment newInstance(AuthFragmentInteraction fragmentInteraction) {
        SignupUserInfoFragment fragment = new SignupUserInfoFragment();
        fragment.fragmentInteraction = fragmentInteraction;
        return fragment;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_signup_user_info, container, false);
        unbinder = ButterKnife.bind(this, view);

        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        if (getArguments() != null) {
            String name = getArguments().getString(KEY_NAME);
            edFirstName.setText(name.substring(0, name.indexOf(" ")));
            edLastName.setText(name.substring(name.lastIndexOf(" ")));
        }

        AuthViewModel viewModel = ViewModelProviders.of(requireActivity()).get(AuthViewModel.class);
        viewModel.getErrors().observe(this, jsonNodes -> {
            if (jsonNodes == null)
                return;
            if (jsonNodes.has("username")) {
                JsonNode userNode = jsonNodes.get("username");
                String msg = userNode.get(0).asText();
                edUsername.setError(msg);
            }
        });

        btnEnter.setOnClickListener(new DebouncedOnClickListener(2000) {
            @Override
            public void onDebouncedClick(View v) {
                onProceedClicked(v);
            }
        });
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        unbinder.unbind();
    }

    @OnClick({R.id.im_female, R.id.im_male})
    public void onGenderIconClicked(View icon) {
        icon.setSelected(true);
        genderChosen = (icon.getId() == R.id.im_male) ? GENDER.MALE : GENDER.FEMALE;
        if (genderChosen == GENDER.MALE) imFemale.setSelected(false);
        else imMale.setSelected(false);
    }

    public void onProceedClicked(View view) {
        String firstname = edFirstName.getText().toString();
        String lastname = edLastName.getText().toString();
        String username = edUsername.getText().toString();
        if (username.isEmpty()) {
            edUsername.setError("This field cannot be empty");
            return;
        }
        if (firstname.isEmpty()) {
            edFirstName.setError("This field cannot be empty");
            return;
        }
        if (genderChosen == null) {
            Toast.makeText(getContext(), "Please select gender", Toast.LENGTH_SHORT).show();
            return;
        }
        fragmentInteraction.onUserInfoProcess(firstname, lastname, username, genderChosen);
    }
}

File: app/src/main/java/com/checkin/app/checkin/Auth/SignupUserInfoFragment.kt
package com.checkin.app.checkin.Auth

import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.Toast
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import butterknife.BindView
import butterknife.OnClick
import com.checkin.app.checkin.R
import com.checkin.app.checkin.misc.fragments.BaseFragment
import com.checkin.app.checkin.user.models.UserModel.GENDER
import com.checkin.app.checkin.utility.DebouncedOnClickListener
import com.checkin.app.checkin.utility.ParentActivityDelegate

class SignupUserInfoFragment : BaseFragment() {
    @BindView(R.id.ed_firstname)
    internal lateinit var edFirstName: EditText
    @BindView(R.id.ed_lastname)
    internal lateinit var edLastName: EditText
    @BindView(R.id.im_male)
    internal lateinit var imMale: FrameLayout
    @BindView(R.id.im_female)
    internal lateinit var imFemale: FrameLayout
    @BindView(R.id.ed_Username)
    internal lateinit var edUsername: EditText
    @BindView(R.id.btn_enter)
    internal lateinit var btnEnter: Button

    private val fragmentInteraction: AuthFragmentInteraction by ParentActivityDelegate(this)
    private var genderChosen: GENDER? = null

    override val rootLayout = R.layout.fragment_signup_user_info

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val name = arguments?.getString(KEY_NAME) ?: ""
        name.indexOf(" ").takeIf { it > 0 }?.let {
            edFirstName.setText(name.substring(0, it))
            edLastName.setText(name.substring(name.lastIndexOf(" ")))
        } ?: kotlin.run {
            edFirstName.setText(name)
        }

        val viewModel = ViewModelProviders.of(requireActivity()).get(AuthViewModel::class.java)
        viewModel.errors.observe(this, Observer {
            it?.let { jsonNodes ->
                if (jsonNodes.has("username")) {
                    val userNode = jsonNodes.get("username")
                    val msg = userNode.get(0).asText()
                    edUsername.error = msg
                }
            }
        })

        btnEnter.setOnClickListener(DebouncedOnClickListener { v ->
            onProceedClicked()
            Unit
        })
    }

    @OnClick(R.id.im_female, R.id.im_male)
    fun onGenderIconClicked(icon: View) {
        icon.isSelected = true
        genderChosen = if (icon.id == R.id.im_male) GENDER.MALE else GENDER.FEMALE
        if (genderChosen == GENDER.MALE)
            imFemale.isSelected = false
        else
            imMale.isSelected = false
    }

    private fun onProceedClicked() {
        val firstname = edFirstName.text.toString()
        val lastname = edLastName.text.toString()
        val username = edUsername.text.toString()
        if (username.isEmpty()) {
            edUsername.error = "This field cannot be empty"
            return
        }
        if (firstname.isEmpty()) {
            edFirstName.error = "This field cannot be empty"
            return
        }
        if (genderChosen == null) {
            Toast.makeText(context, "Please select gender", Toast.LENGTH_SHORT).show()
            return
        }
        fragmentInteraction.onUserInfoProcess(firstname, lastname, username, genderChosen)
    }

    companion object {
        val KEY_NAME = "name"

        fun newInstance() = SignupUserInfoFragment()
    }
}
